processor/src/host/debug.rs

```diff
@@ -1,52 +1,242 @@
+                "Local state before step {} for the context {} at index {}:",
+                self.clk, self.ctx, interval.0
+            )
+        } else {
+            println!(
+                "Local state before step {} for the context {} in the interval [{}, {}]:",
+                self.clk, self.ctx, interval.0, interval.1
+            )
+        };
+
+        print_interval(local_mem_interval, true);
     }
 }
 
 // HELPER FUNCTIONS
 // ================================================================================================
 
-#[cfg(feature = "std")]
-fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
-    // determine how many items to print out
-    let num_items = core::cmp::min(stack.len(), n);
+/// Prints the provided memory interval.
+///
+/// If `is_local` is true, the output addresses are formatted as decimal values, otherwise as hex
+/// strings.
+fn print_interval(mem_interval: Vec<(u32, Option<Word>)>, is_local: bool) {
+    let padding = mem_interval.iter().fold(0, |max, value| word_elem_max_len(value.1).max(max));
 
-    // print all items except for the last one
-    println!("Stack state before step {clk}:");
-    for (i, element) in stack.iter().take(num_items - 1).enumerate() {
-        println!("├── {i:>2}: {element}");
+    // print the main part of the memory (wihtout the last value)
+    for (addr, value) in mem_interval.iter().take(mem_interval.len() - 1) {
+        print_mem_address(*addr, *value, false, is_local, padding)
     }
 
-    // print the last item, and in case the stack has more items, print the total number of
-    // un-printed items
-    let i = num_items - 1;
-    if num_items == stack.len() {
-        println!("└── {i:>2}: {}", stack[i]);
+    // print the last memory value
+    if let Some((addr, value)) = mem_interval.last() {
+        print_mem_address(*addr, *value, true, is_local, padding);
+    }
+}
+
+/// Prints single memory value with its address.
+///
+/// If `is_local` is true, the output address is formatted as decimal value, otherwise as hex
+/// string.
+fn print_mem_address(
+    addr: u32,
+    value: Option<Word>,
+    is_last: bool,
+    is_local: bool,
+    padding: usize,
+) {
+    if let Some(value) = value {
+        if is_last {
+            if is_local {
+                println!(
+                    "└── {addr:>5}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]\n",
+                    value[0].as_int(),
+                    value[1].as_int(),
+                    value[2].as_int(),
+                    value[3].as_int(),
+                    width = padding
+                )
+            } else {
+                println!(
+                    "└── {addr:#010x}: [{:>width$}, {:>width$}, {:>width$}, {:>width$}]\n",
+                    value[0].as_int(),
+                    value[1].as_int(),
+                    value[2].as_int(),
+                    value[3].as_int(),
+                    width = padding
+                )
+            }
```

Comment:
I'd probably do this like so:
```Rust
if is_local {
    print!("└── {addr:>5}: ");
} else {
    print!("└── {addr:#010x}: ");
}
print_word(value, padding);
println!();
```
Where `print_word()` is a new function which uses `println!` to print out the word.